rust/src/daemon.rs

```diff
@@ -420,6 +438,58 @@ pub fn parse_revision(revision: &str) -> CxxResult<ParsedRevision> {
     })
 }
 
+/// Generate refs for baselayer.
+///
+/// For each deployment, if they are layered deployments, then create a ref
+/// pointing to their bases. This is mostly to work around ostree's auto-ref
+/// cleanup. Otherwise we might get into a situation where after the origin ref
+/// is updated, we lose our parent, which means that users can no longer
+/// add/delete packages on that deployment. (They can always just re-pull it, but
+/// let's try to be nice).
+#[context("Generating baselayer refs")]
+pub fn generate_baselayer_refs(
+    mut ffi_sysroot: Pin<&mut crate::ffi::OstreeSysroot>,
+    mut ffi_repo: Pin<&mut crate::ffi::OstreeRepo>,
+    mut ffi_cancellable: Pin<&mut crate::FFIGCancellable>,
+) -> CxxResult<()> {
+    let sysroot = &ffi_sysroot.gobj_wrap();
+    let repo = ffi_repo.gobj_wrap();
+    let cancellable = &ffi_cancellable.gobj_wrap();
+
+    // Delete all the refs.
+    {
+        let refs = repo.list_refs_ext(
+            Some("rpmostree/base"),
+            ostree::RepoListRefsExtFlags::NONE,
+            Some(cancellable),
+        )?;
+
+        for ref_entry in refs.keys() {
+            repo.transaction_set_refspec(ref_entry, None);
+        }
+    }
+
+    // Collect the checksums.
+    let mut bases = vec![];
+    {
+        // Existing deployments.
+        for deployment in sysroot.deployments() {
+            let layered_meta = deployment_layeredmeta_load_commit(&repo, &deployment)?;
+            if layered_meta.is_layered {
+                bases.push(layered_meta.base_commit);
+            }
+        }
+    }
```

Comment:
This is fine as is but I think would be more elegant using combinators, like:

```
let bases: Vec<_> = sysroot.deployments.iter().filter_map(|d| {
   let layered_meta = deployment_layeredmeta_load_commit(&repo, &deployment)?;
   layered_meta.is_layered.then(|| layered_meta.base_commit)
}).collect();
```
or so.

 (not tested)

Reply:
OK and looking below actually since we call `into_iter()` we don't even need to `.collect()` at all and keep it as an iterator.